data_owner_ids.py

```diff
@@ -18,30 +24,99 @@ def parse_args():
     return args
 
 
-def process_csv(csv_path, system_csv_path, system):
-    with open(csv_path) as csvfile:
+def zip_and_clean(system_output_path, system, timestamp):
+    with zipfile.ZipFile(
+        system_output_path.parent / f"{system}.zip", mode="w"
+    ) as system_archive:
+        system_archive.write(
+            system_output_path / f"{system}{timestamp}.csv",
+            Path(system) / f"{system}{timestamp}.csv",
+        )
+        system_archive.write(
+            system_output_path / f"{system}-metadata{timestamp}.json",
+            Path(system) / f"{system}-metadata{timestamp}.json",
+        )
+    print(system_output_path.parent / f"{system}.zip", "created")
+    shutil.rmtree(system_output_path)
+    print("Uncompressed directory removed")
+
+
+def process_output(link_id_path, output_path, system, metadata):
+    data_owner_id_time = datetime.now()
+    timestamp = data_owner_id_time.strftime("%Y%m%dT%H%M%S")
+    n_rows = 0
+    with open(link_id_path) as csvfile:
         reader = csv.DictReader(csvfile)
-        with open(system_csv_path, "w", newline="") as system_csvfile:
+        with open(
+            output_path / f"{output_path.name}{timestamp}.csv", "w", newline=""
+        ) as system_csvfile:
             writer = csv.DictWriter(system_csvfile, fieldnames=["LINK_ID", system])
             writer.writeheader()
             for row in reader:
                 if len(row[system]) > 0:
+                    n_rows += 1
                     writer.writerow({"LINK_ID": row["LINK_ID"], system: row[system]})
+    system_metadata = {
+        "link_id_metadata": {
+            "creation_date": metadata["creation_date"],
+            "uuid1": metadata["uuid1"],
+        },
+        "input_system_metadata": {
+            key: val for key, val in metadata["input_system_metadata"][system].items()
+        },
+        "output_system_metadata": {
+            "creation_date": data_owner_id_time.isoformat(),
+            "number_of_records": n_rows,
+            "uuid1": str(uuid.uuid1()),
+        },
+    }
+    with open(
+        output_path / f"{output_path.name}-metadata{timestamp}.json", "w", newline=""
+    ) as system_metadata_file:
+        json.dump(system_metadata, system_metadata_file, indent=2)
+    return timestamp
 
 
 def do_data_owner_ids(c):
     if c.household_match:
-        csv_path = Path(c.matching_results_folder) / "household_link_ids.csv"
+        link_ids = sorted(
+            Path(c.matching_results_folder).glob("household_link_ids*.csv")
+        )
+    else:
+        link_ids = sorted(Path(c.matching_results_folder).glob("link_ids*.csv"))
+
+    if len(link_ids) > 1:
+        print("More than one link_id file found")
+        link_id_times = [
+            datetime.strptime(
+                x.name.replace("link_ids", "")
+                .replace("household", "")
+                .replace(".csv", ""),
+                "%Y%m%dT%H%M%S",
+            )
+            for x in link_ids
+        ]
```

Comment:
Do you really need to do this if you've already sorted the link ids on line 82 and 86? If not can this be removed?

Reply:
I'm doing this because I need to extract datetime objects from those link_id filenames so that I can run a `max()` on those to grab the latest data (and then the corresponding file). I'm not convinced that this behavior is entirely necessary (though if a linkage agent runs `link_ids.py` multiple times we'll need a way to distinguish between the multiple outputs/a rule that old ones are moved elsewhere), but if we leave this behavior in, we need a way extract the times (though now that I think about it a string slice would do the same thing with a lot more simplicity. Thoughts on that?